[Docs] Guide on setting query context (#18179)
apache/druid: docs/querying/set-query-context.md

Common use cases for the query context include:
- Override default timeouts for long-running queries or complex aggregations.
- Debug query performance by disabling caching during testing.
- Configure SQL-specific behaviors like time zones for accurate time-based analysis.
- Set priorities to ensure critical queries get computational resources first.
- Adjust memory limits for queries that process large datasets.

The way you set the query context depends on how you submit the query to Druid, whether using the web console or API.
It also depends on whether your query is Druid SQL or a JSON-based native query.
This guide shows you how to set the query context for each application.

Before you begin, identify which context parameters you need to configure in order to establish your query context as query context carriers. For available parameters and their descriptions, see [Query context reference](query-context-reference.md).

## Web console

> You can configure query context parameters is via the [Web console](../operations/web-console.md). In the web console, you can set up context parameters for both Druid SQL and native queries.

Author: Updated!

The following steps show you how to set the query context using the web console:

1. In the web console, select **Query** from the top-level navigation.

   ![Query view](../assets/set-query-context-query-view.png)

1. **Click** the **Engine** selector next to the **Run** button to choose the appropriate query type. Unless otherwise instructed, you can leave the engine as `Auto` to let Druid choose the best engine for you.

   ![Engine selection](../assets/set-query-context-select-engine.png)

> 2. Enter the query you want to run.

Author: Updated this section accordingly with new example!

   ```json
   {
     "timeout": 300000,
     "useCache": false
   }
   ```

   The web console validates the JSON object containing the query context parameters and highlights any syntax errors before you click **Run** the query.

   ![Setting the context parameters](../assets/set-query-context-set-context-parameters.png)

6. Click **Run** to execute your query with the specified context parameters.

   ![Running the query](../assets/set-query-context-run-the-query.png)

For more information about using the web console Query view, see [Query view](../operations/web-console.md#query).

## Druid SQL
When using Druid SQL programmatically—such as in applications, automated scripts, or database tools—you can set query context through various methods depending on how your queries are executed. 

### HTTP API

When using the HTTP API, you include query context parameters in the `context` object of your JSON request. For more information on how to format Druid SQL API requests and handle responses, see [Druid SQL API](../api-reference/sql-api.md).

The following example sets the `sqlTimeZone` parameter:

   ```json
   {

>      "query" : "SELECT COUNT(*) FROM data_source WHERE foo = 'bar' AND __time > TIMESTAMP '2000-01-01 00:00:00'",

Author: Updated that too with wikipedia as data source

     "context" : {
       "sqlTimeZone" : "America/Los_Angeles"
     }
   }
   ```

Druid executes your query using the specified context parameters and return the results.

You can set multiple context parameters in a single request:

```json
{
  "query" : "SELECT COUNT(*) FROM data_source WHERE foo = 'bar'",
  "context" : {
    "timeout" : 30000,
    "useCache" : false,
    "sqlTimeZone" : "America/Los_Angeles"
  }
}
```

### JDBC driver API

You can connect to Druid over JDBC and issue Druid SQL queries using this [Druid SQL JDBC driver API](../api-reference/sql-jdbc.md). When connecting to Druid through JDBC, you set query context parameters a JDBC connection properties object. This approach is useful when integrating Druid with BI tools or Java applications.

For example, you can set query context parameters when creating your JDBC connection:

```java

> String url = "jdbc:avatica:remote:url=http://localhost:8082/druid/v2/sql/avatica/";

Reviewer: Where did 8082 come from?

```sql
SET useApproximateTopN = false;
SET sqlTimeZone = 'America/Los_Angeles';
SET timeout = 90000;
SELECT some_column, COUNT(*) 
FROM druid.foo 
WHERE other_column = 'foo' 
GROUP BY 1 
ORDER BY 2 DESC
```

You can also include your SET statements as part of the query string in your HTTP API call. For example,

```bash
curl -X POST 'http://localhost:8888/druid/v2/sql' \
  -H 'Content-Type: application/json' \
  -d '{
    "query": "SET useApproximateTopN = false; SET sqlTimeZone = '\''America/Los_Angeles'\''; SET timeout = 90000; SELECT some_column, COUNT(*) FROM druid.foo WHERE other_column = '\''foo'\'' GROUP BY 1 ORDER BY 2 DESC"
  }'
```

You can also combine SET statements with the `context` field. If you include both, the parameter value in SET takes precedence:

```json
{
  "query": "SET timeout = 90000; SELECT COUNT(*) FROM data_source",
  "context": {
    "timeout": 30000,  // This will be overridden by SET
    "priority": 100    // This will still apply
  }
}
```

For more details on how to use the SET command in your SQL query, see [SET](sql.md#set).

:::info
You cannot use SET statements in JDBC connections.
:::

## Native queries

> For native queries, you can include query context parameters in a JSON object named `context` within your query structure or through [Web Console](./set-query-context.md#web-console).

Author: updated

The following example shows a native query that sets the given query id through context parameters from dataset `wikipedia`:

```json
{
  "queryType": "timeseries",
  "dataSource": "wikipedia",
  "granularity": "day",
  "descending": true,
  "filter": {
    "type": "and",
    "fields": [
      { "type": "selector", "dimension": "countryName", "value": "Australia" },
      { "type": "selector", "dimension": "isAnonymous", "value": "true" }
    ]
  },
  "aggregations": [
    { "type": "count", "name": "row_count" }
  ],
  "intervals": ["2015-09-12T00:00:00.000/2015-09-13T00:00:00.000"],
  "context": {
    "queryId": "only_query_id_test"
  }
}
```

For more information about native queries, see [Native queries](querying.md).

> ## Query Context Precedence

Author: Updated!

When you set query context parameters in Druid, Druid determines which values to use based on the following order of precedence, from lowest to highest:

1. **Built-in hard-coded defaults** — these are the system’s default values used if you don’t specify anything else.
2. **Runtime properties** — if you configure parameters as `druid.query.default.context.{property_key}` in Druid’s configuration files, these override the built-in defaults and act as your system-wide defaults.  For more information, see [Overriding default query context values](../configuration/index.md#overriding-default-query-context-values).

3. **Context parameters you set in your query** — whether in the JSON `context` object or included directly in your queries, these override both the built-in defaults and the runtime properties.

> 4. **SET statements** — when using SET, any parameters you set with `SET key = value;` commands take the highest precedence and override all other settings.

Author: I updated this section accordingly in the newest update

---
id: set-query-context

> title: "Set query context"

Reviewer 2: Suggest a more user-centric title:
Modify query behavior with the query context
Configure query behavior with the query context.

sidebar_label: "Set query context"

> ---

Reviewer 2: Suggest inclduing a `description:` front matter that summarizes the document. Not widely used in the Druid docs right now, but it is helpful with search results.

  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->
  

The query context gives you fine-grained control over how Apache Druid executes your individual queries. While the default settings in Druid work well for most queries, you can set query context to handle specific requirements and optimize performance.

Common use cases for the query context include:
- Override default timeouts for long-running queries or complex aggregations.
- Control resource usage to prevent expensive queries from overwhelming your cluster.
- Debug query performance by disabling caching during testing.
- Configure SQL-specific behaviors like time zones for accurate time-based analysis.
- Set priorities to ensure critical queries get computational resources first.
- Adjust memory limits for queries that process large datasets.

Druid provides several ways to set query context, and the method you use depends on how and where you're submitting your query.

> This guide lists how to set query context for each method of submitting queries.

Reviewer: "how to set **the** query context"

Before you begin, identify which context parameters you need to configure in order to establish your query context as query context carriers. For available parameters and their descriptions, see [Query context reference](query-context-reference.md).

## Druid web console

The most straightforward method to configure query context parameters is via the Druid web console. In web console, you can set up context parameters for both Druid SQL and native queries.

The following steps outline how to define query context parameters:

1. Open the **Query** tab in the web console.

1. **Click** the **Engine** selector next to the **Run** button to choose the appropriate query type:

- Selects the **JSON (native) engine** for native queries.
- Selects the **SQL (native) engine** for Druid SQL queries.
- Selects the **SQL (task) engine** for Multi-stage queries (MSQ).

> - Selects the **Auto engine** to let the console detect the query type automatically. You just paste your query into the **Query** view, and the web console chooses the right engine for you.

Reviewer 2: +1 for @vtlim 's comment here since the DART engine is coming. Maybe it is not worth maintaining a list of engines here and just say. "Unless otherwsise instructed, you can leave the engine as `auto` to let Druid choose the best engine for you.

2. Enter the query you want to run in the web console.

3. Select **Edit context** from the menu.
4. In the **Edit query context** dialog, add your context parameters as JSON key-value pairs:
   ```json
   {
     "timeout": 300000,
     "useCache": false
   }
   ```
5. Click **Save** to apply the context to your query.
6. Click **Run** to execute your query with the specified context parameters.

The web console validates your JSON and highlights any syntax errors before you run the query.

For more information about using the Druid SQL Web Console Query view, see [Query view](../operations/web-console.md#query).

## Druid SQL

> When you use Druid SQL programmatically—such as in applications, automated scripts, or database tools, you can set query context using three different methods depending on how you execute your queries beyond [Druid Web Console](./set-query-context.md#druid-web-console). 

Reviewer: Druid Web Console -- fix casing. Also check how this is referred to in other places. Is it called "Druid web console" everywhere or just "web console"?

> When you use Druid SQL programmatically—such as in applications, automated scripts, or database tools, you can set query context using three different methods depending on how you execute your queries beyond [Druid Web Console](./set-query-context.md#druid-web-console). 

Reviewer 2: +1 to @vtlim comment, also wondering if it is necessary to mention that these are "beyond the [W/w]eb consloe"

### HTTP API

When using the HTTP API, you include query context parameters in the `context` object of your JSON request. 

The following example sets the `sqlTimeZone` parameter:

   ```json
   {
     "query" : "SELECT COUNT(*) FROM data_source WHERE foo = 'bar' AND __time > TIMESTAMP '2000-01-01 00:00:00'",
     "context" : {
       "sqlTimeZone" : "America/Los_Angeles"
     }
   }
   ```

> Druid will execute your query with the specified context parameters and return the results.

Reviewer 2: ```suggestion
Druid executes your query using the specified context parameters and return the results.
```

### JDBC driver API

When connecting to Druid through JDBC, you set query context parameters a JDBC connection properties object. This approach is useful when integrating Druid with BI tools or Java applications.

Druid uses the Avatica JDBC driver (version 1.23.0 or later recommended). Note that Avatica does not support passing connection context parameters from the JDBC connection string—you must use a `Properties` object instead.

You can set query context parameters when creating your JDBC connection:

```java
String url = "jdbc:avatica:remote:url=http://localhost:8082/druid/v2/sql/avatica/";

// Set any query context parameters you need here.
Properties connectionProperties = new Properties();
connectionProperties.setProperty("sqlTimeZone", "America/Los_Angeles");
connectionProperties.setProperty("useCache", "false");

try (Connection connection = DriverManager.getConnection(url, connectionProperties)) {
  // create and execute statements, process result sets, etc
}
```

For more details on how to use JDBC driver API, see [Druid SQL JDBC driver API](../api-reference/sql-jdbc.md).

### SET statements

> Beyond using the `context` parameter, you can use `SET` command to specify SQL query context parameters that modify the behavior of a Druid SQL query. You can include one or more SET statements before the main SQL query. You can use `SET` in the both web console and Druid SQL HTTP API. 

Reviewer 2: Three senteces in a row with "you can" construction. Vary the structure for example:

You can use the `SET` command to specify SQL query context parameters that modify the behavior of a Druid SQL query. Druid accepts one or more SET statements before the main SQL query. The `SET` command works in the both web console and the Druid SQL HTTP API.

WHERE other_column = 'foo' 
GROUP BY 1 
ORDER BY 2 DESC
```

You can also include your SET statements as part of the query string in your HTTP API call. For example,

```bash
curl -X POST 'http://localhost:8888/druid/v2/sql' \
  -H 'Content-Type: application/json' \
  -d '{
    "query": "SET useApproximateTopN = false; SET sqlTimeZone = '\''America/Los_Angeles'\''; SET timeout = 90000; SELECT some_column, COUNT(*) FROM druid.foo WHERE other_column = '\''foo'\'' GROUP BY 1 ORDER BY 2 DESC"
  }'
```

You can also combine SET statements with the `context` field. If you include both, the parameter value in SET takes precedence:

```json
{
  "query": "SET timeout = 90000; SELECT COUNT(*) FROM data_source",
  "context": {
    "timeout": 30000,  // This will be overridden by SET
    "priority": 100    // This will still apply
  }
}
```

SET statements only apply to the query in the same request. Subsequent requests are not affected. 

SET statements work with SELECT, INSERT, and REPLACE queries.

For more details on how to use the SET command in your SQL query, see [SET](sql.md#set).

:::info
 You cannot use SET statements when using Druid SQL JDBC connections.
:::

## Native queries

> For native queries, you can include query context parameters in a JSON object named `context` within your query structure or through [Druid Web Console](./set-query-context.md#druid-web-console).

Reviewer: See earlier comment on web console

```sql
SET useApproximateTopN = false;
SET sqlTimeZone = 'America/Los_Angeles';
SET timeout = 90000;
SELECT some_column, COUNT(*) 
FROM druid.foo 
WHERE other_column = 'foo' 
GROUP BY 1 
ORDER BY 2 DESC
```

You can also include your SET statements as part of the query string in your HTTP API call. For example,

```bash
curl -X POST 'http://localhost:8888/druid/v2/sql' \
  -H 'Content-Type: application/json' \
  -d '{
    "query": "SET useApproximateTopN = false; SET sqlTimeZone = '\''America/Los_Angeles'\''; SET timeout = 90000; SELECT some_column, COUNT(*) FROM druid.foo WHERE other_column = '\''foo'\'' GROUP BY 1 ORDER BY 2 DESC"
  }'
```

You can also combine SET statements with the `context` field. If you include both, the parameter value in SET takes precedence:

```json
{
  "query": "SET timeout = 90000; SELECT COUNT(*) FROM data_source",
  "context": {
    "timeout": 30000,  // This will be overridden by SET
    "priority": 100    // This will still apply
  }
}
```

For more details on how to use the SET command in your SQL query, see [SET](sql.md#set).

:::info
You cannot use SET statements in JDBC connections.
:::

## Native queries

> For native queries, you can include query context parameters in a JSON object named `context` within your query structure or through [Web Console](./set-query-context.md#web-console).

Author: ```suggestion
For native queries, you can include query context parameters in a JSON object named `context` within your query or through [web console](./set-query-context.md#web-console).
```

---
id: set-query-context
title: "Set query context"
sidebar_label: "Set query context"
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ "License"); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->
  

> Query context gives you fine-grained control over how Apache Druid executes your individual queries. While Druid's default settings work well for most queries, you can set query context to handle specific requirements and optimize performance.

Reviewer: ```suggestion
The query context gives you fine-grained control over how Apache Druid executes your individual queries. While the default settings in Druid work well for most queries, you can set the query context to handle specific requirements and optimize performance.
```

> You'll use query context when you need to:

Reviewer: * Remove future tense
* Think about rewording the intro sentence to let the reader know that these are examples (not a complete list)

- Override default timeouts for long-running queries or complex aggregations.
- Control resource usage to prevent expensive queries from overwhelming your cluster.
- Debug query performance by disabling caching during testing.
- Configure SQL-specific behaviors like time zones for accurate time-based analysis.
- Set priorities to ensure critical queries get computational resources first.
- Adjust memory limits for queries that process large datasets.

> Druid offers multiple approaches for you to set up your query context. This guide covers all the approaches—from the web console to programmatic APIs with practical examples you can adapt for your use case.

Reviewer: Consider rephrasing this section, since there's not really multiple approaches by the way you send your query, but rather the way you send your query has 1-2 options. Suggest making it more general.

Before you begin, identify which context parameters you need to configure in order to establish your query context as context carriers. For a complete list of available parameters and their descriptions, see [Query context reference](query-context-reference.md).

> ## Druid web console

Reviewer: We should test/clarify if setting query context parameters here apply to both Druid SQL _and_ native queries

The easiest way to set query context is through the Druid web console. You can define context parameters using the Query context editor with these steps:

> 1. Open the **Query** tab in the Web Console.

Reviewer: Check capitalization!

The web console validates your JSON and highlights any syntax errors before you run the query.

For more information about using the Druid SQL Web Console Query view, see [Query view](../operations/web-console.md#query).

## Druid SQL

When you use Druid SQL programmatically—such as in applications, automated scripts, or database tools, you can set query context using three different methods depending on how you execute your queries beyond web console. 

### HTTP API

When using the HTTP API, you include query context parameters in the `context` object of your JSON request. 

To set query context via the HTTP API, You can start with your SQL query:

   ```sql
   SELECT COUNT(*) FROM data_source WHERE foo = 'bar' AND __time > TIMESTAMP '2000-01-01 00:00:00'
   ```

Then you need to create a JSON request with your query and context parameters. The following example sets the `sqlTimeZone` parameter:

   ```json
   {
     "query" : "SELECT COUNT(*) FROM data_source WHERE foo = 'bar' AND __time > TIMESTAMP '2000-01-01 00:00:00'",
     "context" : {
       "sqlTimeZone" : "America/Los_Angeles"
     }
   }
   ```

> You can then send the request to Druid's SQL API HTTP endpoint:

Reviewer: suggest removing

**Basic example:**

You can set query context parameters when creating your JDBC connection:

```java
// Connect to /druid/v2/sql/avatica/ on your Broker
String url = "jdbc:avatica:remote:url=http://localhost:8888/druid/v2/sql/avatica/;transparent_reconnection=true";

// Set any query context parameters you need here
Properties connectionProperties = new Properties();
connectionProperties.setProperty("sqlTimeZone", "America/Los_Angeles");
connectionProperties.setProperty("useCache", "false");
connectionProperties.setProperty("priority", "100");

try (Connection connection = DriverManager.getConnection(url, connectionProperties)) {
    try (
        final Statement statement = connection.createStatement();
        final ResultSet resultSet = statement.executeQuery("SELECT COUNT(*) FROM data_source WHERE foo = 'bar'")
    ) {
        while (resultSet.next()) {
            // Process result set
            System.out.println("Count: " + resultSet.getLong(1));
        }
    }
}
```

> **With authentication:**

Reviewer: Suggest removing this if it's already covered; if you want, you can mention it in a line and link to the specific spot

      { "type": "or",
        "fields": [
          { "type": "selector", "dimension": "sample_dimension2", "value": "sample_value2" },
          { "type": "selector", "dimension": "sample_dimension3", "value": "sample_value3" }
        ]
      }
    ]
  },
  "aggregations": [
    { "type": "longSum", "name": "sample_name1", "fieldName": "sample_fieldName1" },
    { "type": "doubleSum", "name": "sample_name2", "fieldName": "sample_fieldName2" }
  ],
  "postAggregations": [
    { "type": "arithmetic",
      "name": "sample_divide",
      "fn": "/",
      "fields": [
        { "type": "fieldAccess", "name": "postAgg__sample_name1", "fieldName": "sample_name1" },
        { "type": "fieldAccess", "name": "postAgg__sample_name2", "fieldName": "sample_name2" }
      ]
    }
  ],
  "intervals": [ "2012-01-01T00:00:00.000/2012-01-03T00:00:00.000" ],
  // Add context parameters here
  "context": {
    "timeout": 30000,        // Query timeout in milliseconds
    "priority": 100,         // Higher priority queries get more resources
    "useCache": false,       // Disable cache for testing
    "finalize": true         // Finalize aggregation results
  }
}
```

> Then you can submit the query using curl:

Reviewer: suggest removing

```bash
# Replace localhost:8888 with your Druid Broker or Router address
curl -X POST 'http://localhost:8888/druid/v2/?pretty' \
  -H 'Content-Type:application/json' \
  -H 'Accept:application/json' \
  -d @query.json
```

> **Using the web console:**

Reviewer: Consolidate with earlier web console section (don't create a separate section here), although you can link to it. Are you able to set parameters in the **Edit context** option in the web console? What happens if you set parameters in both areas (in the console and in the query itself).

Notice that this section predominantly describes "how to send native queries in the web console" which is not in scope of this doc.
